File: 40-combination-sum-ii/40-combination-sum-ii.cpp
class Solution {
public:
    void func(int i, int target, vector<int> &ds, vector<vector<int>> &ans, vector<int> &arr){
        if(target == 0){ 
            ans.push_back(ds);
            return;
        }
        
        for(int ind=i; ind<arr.size(); ind++){
            
            //agar prev element same h to aage badho
            if(ind>i && arr[ind] == arr[ind-1]) continue;
            
            //agar ye number khud bada hua to usse aage ke bhi bade honge... 
            //so no need to check
            if(arr[ind] > target) break;
            
            ds.push_back(arr[ind]);
            func(ind+1, target - arr[ind], ds, ans, arr);
            ds.pop_back();
        }
    }
    
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<int> ds;
        vector<vector<int>> ans;
        
        func(0, target, ds, ans, candidates);
        
        return ans;
    }
};

File: 40-combination-sum-ii/40-combination-sum-ii.py
class Solution:
    def helper(self, i, n, target, nums, subset, ans):
        
        if target == 0: 
            ans.append(subset[:])
            return
        
        for ind in range(i, n):
            #same value ignore
            if(ind > i and nums[ind] == nums[ind-1]): continue
            
            #ye value badi h to target nahi bana payegi...
            #and uske aage ke bhi bade honge
            if(nums[ind] > target): break
            
            subset.append(nums[ind])
            self.helper(ind+1, n, target-nums[ind], nums, subset, ans)
            subset.pop()

    
    
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        n = len(candidates)
        ans = []
        subset = []
        self.helper(0, n, target, candidates, subset, ans)
        return ans